pkg/buses/message_dispatcher_test.go

```diff
@@ -0,0 +1,344 @@
+					"ce-abc":       {"ce-abc-value"},
+				},
+				body: "destination",
+			},
+			fakeResponse: &http.Response{
+				StatusCode: http.StatusAccepted,
+				Header: map[string][]string{
+					"do-not-passthrough": {"no"},
+					"x-request-id":       {"altered-id"},
+					"knative-1":          {"new-knative-1-value"},
+					"ce-abc":             {"new-ce-abc-value"},
+				},
+				Body: ioutil.NopCloser(bytes.NewBufferString("")),
+			},
+		},
+		"destination and reply": {
+			destination: "test-destination-svc.test-namespace.svc.cluster.local",
+			replyTo:     "test-reply-svc.test-namespace.svc.cluster.local",
+			message: &Message{
+				Headers: map[string]string{
+					// do-not-forward should not get forwarded.
+					"do-not-forward": "header",
+					"x-request-id":   "id123",
+					"knative-1":      "knative-1-value",
+					"knative-2":      "knative-2-value",
+					"ce-abc":         "ce-abc-value",
+				},
+				Payload: []byte("destination"),
+			},
+			expectedDestRequest: &requestValidation{
+				url: "http://test-destination-svc.test-namespace.svc.cluster.local/",
+				headers: map[string][]string{
+					"x-request-id": {"id123"},
+					"knative-1":    {"knative-1-value"},
+					"knative-2":    {"knative-2-value"},
+					"ce-abc":       {"ce-abc-value"},
+				},
+				body: "destination",
+			},
+			fakeResponse: &http.Response{
+				StatusCode: http.StatusAccepted,
+				Header: map[string][]string{
+					"do-not-passthrough": {"no"},
+					"x-request-id":       {"altered-id"},
+					"knative-1":          {"new-knative-1-value"},
+					"ce-abc":             {"new-ce-abc-value"},
+				},
+				Body: ioutil.NopCloser(bytes.NewBufferString("destination-response")),
+			},
+			expectedReplyRequest: &requestValidation{
+				url: "http://test-reply-svc.test-namespace.svc.cluster.local/",
+				headers: map[string][]string{
+					"x-request-id": {"altered-id"},
+					"knative-1":    {"new-knative-1-value"},
+					"ce-abc":       {"new-ce-abc-value"},
+				},
+				body: "destination-response",
+			},
+		},
+	}
+	for n, tc := range testCases {
+		t.Run(n, func(t *testing.T) {
+			md := NewMessageDispatcher(zap.NewNop().Sugar())
+			fc := &fakeHttpClient{
+				response: tc.fakeResponse,
+			}
+			md.httpClient = fc
+			err := md.DispatchMessage(tc.message, tc.destination, tc.replyTo, DispatchDefaults{})
+			if tc.expectedErr != (err != nil) {
+				t.Errorf("Unexpected error from DispatchRequest. Expected %v. Actual: %v", tc.expectedErr, err)
+			}
+			if tc.expectedDestRequest != nil {
+				rv := fc.popRequest(t)
+				assertEquality(t, *tc.expectedDestRequest, rv)
+			}
+			if tc.expectedReplyRequest != nil {
+				rv := fc.popRequest(t)
+				assertEquality(t, *tc.expectedReplyRequest, rv)
+			}
+			if len(fc.requests) != 0 {
+				t.Errorf("Unexpected requests: %+v", fc.requests)
+			}
+		})
+	}
+}
+
+type requestValidation struct {
+	url     string
+	headers http.Header
+	body    string
+}
+
+type fakeHttpClient struct {
```

Comment:
The other way seems to be this: https://golang.org/pkg/net/http/httptest/

Reply:
I think the suggestion is to make this fakeHandler. Then use a real http.Client{} and point everything at the in-memory server, using the fakeHandler. Is that correct?

Reply:
Done.